Index: src/lib/SSDP/NotifyMsgFactory.cpp
===================================================================
--- src/lib/SSDP/NotifyMsgFactory.cpp	(revision 692)
+++ src/lib/SSDP/NotifyMsgFactory.cpp	(working copy)
@@ -56,7 +56,7 @@
 	
 	result << "NOTIFY * HTTP/1.1\r\n";
   result << "HOST: 239.255.255.250:1900\r\n";
-	result << "CACHE-CONTROL: max-age=1800\r\n";
+	result << "CACHE-CONTROL: max-age=1900\r\n";
 	result << "LOCATION: http://" << m_sHTTPServerURL << "/description.xml\r\n";
   result << "NT: " << type_to_string(a_type) << "\r\n";
 	result << "NTS: ssdp:alive\r\n";
