--- ata/chipsets/ata-via.c.orig	2016-03-13 20:39:16.000000000 +0100
+++ ata/chipsets/ata-via.c	2016-03-13 20:41:48.000000000 +0100
@@ -150,6 +150,12 @@
     if (ata_setup_interrupt(dev, ata_generic_intr))
 	return ENXIO;
 
+    /* VT6421/6420 fix from http://thread.gmane.org/gmane.linux.kernel/1062139 */
+    if (pci_get_devid(dev) == ATA_VIA6421 || pci_get_devid(dev) == ATA_VIA6420) {
+	pci_write_config(dev, 0x52,
+	    (pci_read_config(dev, 0x52, 1) | 0x04), 1);
+    }
+
     /* 2 SATA with "SATA registers" at PCI config space + PATA on secondary */
     if (ctlr->chip->cfg2 & VIASATA) {
 	ctlr->ch_attach = ata_via_sata_ch_attach;
